kernel/kernel-api/src/main/java/io/delta/kernel/coordinatedcommits/CommitCoordinatorClient.java

```diff
@@ -0,0 +1,168 @@
+ *
+ * <ul>
+ *   <li>{@link #registerTable}: Determine the table config during commit coordinator registration.
+ *   <li>{@link #commit}: Commit a new version of the table.
+ *   <li>{@link #getCommits}: Tracks and returns unbackfilled commits.
+ *   <li>{@link #backfillToVersion}: Ensure that commits are backfilled if/when needed.
+ * </ul>
+ *
+ * @since 3.3.0
+ */
+@Evolving
+public interface CommitCoordinatorClient {
+
+  /**
+   * Register the table represented by the given {@code logPath} at the provided {@code
+   * currentVersion} with the commit coordinator this commit coordinator client represents.
+   *
+   * <p>This API is called when the table is being converted from an existing file system table to a
+   * coordinated-commit table.
+   *
+   * <p>When a new coordinated-commit table is being created, the {@code currentVersion} will be -1
+   * and the upgrade commit needs to be a file system commit which will write the backfilled file
+   * directly.
+   *
+   * @param engine The {@link Engine} instance to use.
+   * @param logPath The path to the delta log of the table that should be converted.
+   * @param tableIdentifier The table identifier for the table, or {@code null} if the table doesn't
+   *     use any identifier (i.e. it is path-based).
+   * @param currentVersion The version of the table just before conversion. currentVersion + 1
+   *     represents the commit that will do the conversion. This must be backfilled atomically.
+   *     currentVersion + 2 represents the first commit after conversion. This will go through the
+   *     CommitCoordinatorClient and the client is free to choose when it wants to backfill this
+   *     commit.
+   * @param currentMetadata The metadata of the table at currentVersion
+   * @param currentProtocol The protocol of the table at currentVersion
+   * @return A map of key-value pairs which is issued by the commit coordinator to uniquely identify
+   *     the table. This should be stored in the table's metadata for table property {@link
+   *     io.delta.kernel.internal.TableConfig#COORDINATED_COMMITS_TABLE_CONF}. This information
+   *     needs to be passed to the {@link #commit}, {@link #getCommits}, and {@link
+   *     #backfillToVersion} APIs to identify the table.
+   */
+  Map<String, String> registerTable(
+      Engine engine,
+      String logPath,
+      @Nullable TableIdentifier tableIdentifier,
+      long currentVersion,
+      AbstractMetadata currentMetadata,
+      AbstractProtocol currentProtocol);
+
+  /**
+   * Commit the given set of actions to the table represented by {@code tableDescriptor}.
+   *
+   * @param engine The {@link Engine} instance to use.
+   * @param tableDescriptor The descriptor for the table.
+   * @param commitVersion The version of the commit that is being committed.
+   * @param actions The set of actions to be committed
+   * @param updatedActions Additional information for the commit, including:
```

Comment:
Sorry a bunch of questions about CC writes not necessarily specific to this PR. What are the `updatedActions` for and why do they need to be separated from the other `actions`?

Reply:
The updated actions are the CommitInfo and the previous and current Metadata/Protocol. They are also included in the `actions` (Protocol and Metadata only if they changed) but we want to pass them separately for convenience in case commit coordinator client implementations want to do something with them (for example check if the Metadata of a table has changed).

Reply:
Got it thanks!

Reply:
Also the `actions` is an iterator i.e. it can only be traversed once. The commit coordinator can't go through it to get these important updates (e.g. schema change / protocol change). So the API explicitly passes such updates.